app/src/main/java/com/github/sdpcoachme/location/provider/FusedLocationProvider.kt

```diff
@@ -0,0 +1,184 @@
+            ActivityResultContracts.RequestPermission()
+        ) { isGranted ->
+            if (isGranted) {
+                checkLocationSetting()
+            } else {
+                setLocationToAddress()
+            }
+        }
+
+        requestSettingLauncher = appContext.registerForActivityResult(
+            ActivityResultContracts.StartIntentSenderForResult()
+        ) {
+            if (it.resultCode == RESULT_OK) {
+                // User enabled device location, we can now retrieve it
+                getDeviceLocation(0)
+            } else {
+                // User did not enable device location
+                setLocationToAddress()
+            }
+        }
+    }
+
+    override fun requestPermission() {
+        requestPermissionLauncher.launch(ACCESS_FINE_LOCATION)
+    }
+
+    /**
+     * Especially useful after application installation.
+     */
+    override fun locationIsPermitted(): Boolean {
+        val locationPermission = ContextCompat.checkSelfPermission(appContext, ACCESS_FINE_LOCATION)
+        return locationPermission == PERMISSION_GRANTED
+    }
+
+    /**
+     * This function is always called with granted permissions
+     * and the device location settings enabled. This safely allows the recursive call needed in the
+     * case the device location service takes time to be deployed. A max iteration "timeout" has
+     * been set to ensure termination.
+     *
+     * Notice that the fusedLocationProviderClient.lastLocation task is null if the location is
+     * disabled on the device (also even if the last location was previously retrieved because
+     * disabling the device location clears the cache).
+     */
+    @SuppressLint("MissingPermission") //permission is checked before the call
+    private fun getDeviceLocation(delay: Long) {
+        // getDeviceLocation should not be called more than MAX_LOCATION_DELAY time
+        if (delay >= DELAY) {
+            error("getDeviceLocation has reached its max recursive delay")
+        }
+        try {
+            fusedLocationProviderClient.lastLocation.addOnCompleteListener {
+                if (it.isSuccessful) {
+                    if (it.result != null) {
+                        lastUserLocation.value = LatLng(it.result.latitude, it.result.longitude)
+                    } else {
+                        // Yann: I haven't found any better way to handle this so I'm open to
+                        // suggestions ! Futures and proper TimeOuts were tested but the
+                        // lastLocation task would not complete successfully nor could I find any
+                        // suitable implementation.
+                        getDeviceLocation(delay + 1)
+                    }
+                }
+            }
+        } catch (e: SecurityException) {
+            error("getDeviceLocation was called without correct permissions : ${e.message}")
+        }
+    }
```

Comment:
I see the issue here... Here is the solution I thought about:

- Create a future eg. f.
- Lauch a coroutine (Globalscope.launch...) with a while loop testing if we successfully get the location. Complete the future once successful.
- Have the rest of the operations done asynchronously for example with thenApply

Hope this can help.

Reply:
Yes this may be a way to implement this. The problem of the while loop is that we cannot catch the completion of the setting enabling so in the end, it would be an arbitrary TimeOut. In fact, this would replace the same behavior as using a manually set maximum recurive calls DELAY.